aptos-move/move-examples/marketplace/sources/listing.move

```diff
@@ -0,0 +1,259 @@
+    /// purchase to extract the token.
+    public entry fun extract_tokenv1(
+        owner: &signer,
+        object: Object<TokenV1Container>,
+    ) acquires TokenV1Container {
+        let object_addr = object::object_address(&object);
+        assert!(object::is_owner(object, signer::address_of(owner)), 0);
+        let TokenV1Container {
+            token,
+            delete_ref: _,
+            transfer_ref: _,
+        } = move_from(object_addr);
+        tokenv1::deposit_token(owner, token);
+    }
+
+    /// The listing has concluded, transfer the asset and delete the listing. Returns the seller
+    /// for depositing any profit and the fee schedule for the marketplaces commission.
+    public(friend) fun complete(
+        object: Object<Listing>,
+        recipient: address,
+    ): (address, Object<FeeSchedule>) acquires Listing, TokenV1Container {
+        let listing_addr = object::object_address(&object);
+        let Listing {
+            object,
+            seller,
+            fee_schedule,
+            start_time: _,
+            delete_ref,
+            extend_ref,
+        } = move_from<Listing>(listing_addr);
+
+        let obj_signer = object::generate_signer_for_extending(&extend_ref);
+        if (exists<TokenV1Container>(object::object_address(&object))) {
+            extract_or_transfer_tokenv1(recipient, object::convert(object));
+        } else {
+            object::transfer(&obj_signer, object, recipient);
+        };
+        object::delete(delete_ref);
+
+        (seller, fee_schedule)
+    }
+
+    public(friend) fun assert_started(object: &Object<Listing>): address acquires Listing {
+        let listing_addr = object::object_address(object);
+        assert!(exists<Listing>(listing_addr), error::not_found(ENO_LISTING));
+
+        let listing = borrow_global<Listing>(listing_addr);
+        let now = timestamp::now_seconds();
+        assert!(listing.start_time <= now, error::invalid_state(ELISTING_NOT_STARTED));
+        listing_addr
+    }
+
+    // View
+
+    #[view]
+    public fun seller(object: Object<Listing>): address acquires Listing {
+        let listing = borrow_listing(object);
+        listing.seller
+    }
+
+    #[view]
+    public fun listed_object(object: Object<Listing>): Object<ObjectCore> acquires Listing {
+        let listing = borrow_listing(object);
+        listing.object
+    }
+
+    #[view]
+    public fun fee_schedule(object: Object<Listing>): Object<FeeSchedule> acquires Listing {
+        let listing = borrow_listing(object);
+        listing.fee_schedule
+    }
+
+    #[view]
+    public fun compute_royalty(
+        object: Object<Listing>,
+        amount: u64,
+    ): (address, u64) acquires Listing, TokenV1Container {
+        let listing = borrow_listing(object);
+        let obj_addr = object::object_address(&listing.object);
+        if (exists<TokenV1Container>(obj_addr)) {
+            let token_container = borrow_global<TokenV1Container>(obj_addr);
```

Comment:
This math can be almost one-lined by creating a FixedPoint64 from numerator/denom and using the math libraries provided there

Reply:
I feel fixedpoint is too heavy... as it converts everything to u128 and then u256, but agreed we can make this safer.

Reply:
You can also use https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/math64.move#L32